freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/FmsUser.java

```diff
@@ -10,6 +10,8 @@
 
     private String lastName;
 
+    private WorkloadCredential creds;
```

Comment:
we need to be mindful of how changes to the FmsUser will affect state difference calculations. I see that the creds are not added to the `equals` or `hashCode` methods. This will currently provide the correct behavior (since we use equality to compare against the ipa state and the ipa state won't include workload credentials) but this breaks equality between FmsUsers.
I think it would be better to update `equals` and `hashCode` here and perhaps move the `UsersState` models to include a map from name-->FmsUser rather than a set. We can update the difference code to use filters rather than `Sets.difference`

Reply:
So, I looked into the equals method of FmsUser also, hashcode only consider username, and both names so that should not be an issue. Also methods like such:  getUsersToBeRemoved(Collection<String> umsStateUsers, Collection<String> ipaStateUsers) are String based collections. 

But, current implementation equals method and hashcode doesn't seems right. We are checking extra information like first name and last name. We should only check for workload username because that's the unique information in IPA and any two or more objects having the same workload username (username) must be considered as similar. Consider the scenario where same workloadusername obj csso_apurohit's first name is changed in IPA manually (however we don't recommend (allow) that but it is possible), in that case, the user is same (since same workload username) hashcode could be different.

To avoid above, I can add a new Map entry in UserState for user -> Wrokload Creds. Fow now that should be good.

Reply:
I have moved this to UserState and as we discussed, we should re-visit and go more as interfaces and child classes. Fix related to this can be independent of this PR.